### ISSUE TITLE
Repeatedly calling "all-types" command downgrades performance

### ISSUE THREAD

**Issue by Trebor-Huang** [dramforever/vscode-ghc-simple#45]
## Status

- [ ] A workaround is available
- [ ] A fix is written
- [ ] A fix is released
- [ ] The fixed verison is on the marketplace.

---

- Version of vscode-ghc-simple: v0.1.13
- Version of VSCode: 1.40.2
- Version of Haskell-related tools:
    
    - GHC: 8.4.4
    - Stack: 2.1.3 (brew installation)
    - Cabal: None

## Steps to reproduce

1. Open a relatively big haskell project.
2. Do some edit.
3. Save the file.
4. Place the cursor at some expression to trigger the range-type functionality.
5. Repeat 2, 3 and 4.

Each time this causes a lag in the editor, since a minor edit-save causes the type cache to be cleared and re-cached. In a moderately large project (~500 lines, 5 files) the impact on performance is already significant. The `:all-types` command produces 2000+ lines of information.

My suggestion is that the plugin stop calling the `:all-types` command when the project grows to a certain size, and caches only the type information that is asked for.

**Comment by dramforever** [comment]
You have correctly noticed that `:all-types` is vastly suboptimal.

Ideally, if I can get GHC to dump from the parser where everything is, I can just find where the selection lies, and just ask `:type-at`. Or maybe I can get the all the type information in a more efficient way, since it seems that GHC *is* able to collect them rather quickly. After searches through the documentation I could not find anything better than `:all-types`. `-ddump-parsed-ast` seems really close, but it isn't really friendly to parse and I can't seem to figure out a way 'get' it (for example, debugging flags cannot be turned back off, so it messes with other stuff)...

I guess this is partly where vscode-ghc-simple fundamentally differs from say HIE. For example, I don't really want to interface with the GHC API. The entire project started because GHCi provided features like `:type-at`, and I kinda want to stick to what GHCi provides us. (In this regards it's similar to [dante](https://github.com/jyp/dante), which I learned about after starting this.)

So basically we are stranded.

One idea is that we ditch being able to 'intelligently' detect which expression the user's selection is about, and instead pick the word under cursor or exactly the selection. However last time I checked `:type-at` goes really wacky if you don't send it an actual existing expression, so this way we will be showing the user wrong types, which really might hurt more than it helps.

**Comment by dramforever** [comment]
> My suggestion is that the plugin stop calling the `:all-types` command when the project grows to a certain size, and caches only the type information that is asked for.

I guess I didn't address this directly. Doing this would mean that for a large project, every time you edit a file, you lose your ability to see types for this file, which is actually not far from having range type disabled altogether.

Unrelatedly, I'm closing as a duplicate of #39.
